Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -356,6 +360,205 @@ alias ⟨IsIndContained.exists_iso_subgraph, IsIndContained.of_exists_iso_subgra
 @[simp] lemma compl_isIndContained_compl : Gᶜ ⊴ Hᶜ ↔ G ⊴ H :=
   Embedding.complEquiv.symm.nonempty_congr
 
-protected alias ⟨IsIndContained.of_compl, IsIndContained.compl⟩ := compl_isIndContained_compl
+/-!
+### Counting the copies
+
+If `G` and `H` are finite graphs, we can count the number of unlabelled and labelled copies of `G`
+in `H`.
+
+#### Not necessarily induced copies
+-/
+
+section CopyCount
+variable [Fintype V]
+
+/-- `G.copyCount H` is the number of unlabelled copies of `H` in `G`.
+See `SimpleGraph.labelledCopyCount` for the number of labelled copies. -/
+noncomputable def copyCount (G : SimpleGraph V) (H : SimpleGraph W) : ℕ := by
+  classical exact #{G' : G.Subgraph | Nonempty (H ≃g G'.coe)}
+
+@[simp] lemma copyCount_bot (G : SimpleGraph V) : copyCount G (⊥ : SimpleGraph V) = 1 := by
+  classical
+  rw [copyCount]
+  convert card_singleton (α :=  G.Subgraph)
+    { verts := .univ
+      Adj := ⊥
+      adj_sub := False.elim
+      edge_vert := False.elim }
+  simp only [eq_singleton_iff_unique_mem, mem_filter, mem_univ, Subgraph.coe_bot, true_and,
+    Nonempty.forall]
+  refine ⟨⟨⟨(Equiv.Set.univ _).symm, by simp⟩⟩, fun H' e ↦
+    Subgraph.ext ((set_fintype_card_eq_univ_iff _).1 <| Fintype.card_congr e.toEquiv.symm) ?_⟩
+  ext a b
+  simp only [Prop.bot_eq_false, Pi.bot_apply, iff_false]
+  exact fun hab ↦ e.symm.map_rel_iff.2 hab.coe
+
+@[simp] lemma copyCount_of_isEmpty [IsEmpty W] (G : SimpleGraph V) (H : SimpleGraph W) :
+    G.copyCount H = 1 := by
+  rw [copyCount]
+  convert card_singleton (⊥ : G.Subgraph)
+  simp only [eq_singleton_iff_unique_mem, mem_filter, mem_univ, Subgraph.coe_bot, true_and,
+    Nonempty.forall, Subsingleton.elim H ⊥]
+  haveI : IsEmpty (⊥ : G.Subgraph).verts := by simp
+  refine ⟨⟨⟨⟨isEmptyElim, isEmptyElim, isEmptyElim, isEmptyElim⟩, fun {a} ↦ isEmptyElim a⟩⟩,
+    fun H' e ↦ Subgraph.ext ?_ <| funext₂ fun a b ↦ ?_⟩
+  · simpa [Set.eq_empty_iff_forall_not_mem, filter_eq_empty_iff, ‹IsEmpty W›] using
+      e.toEquiv.symm.isEmpty_congr
+  · simp only [Subgraph.not_bot_adj, eq_iff_iff, iff_false]
+    exact fun hab ↦ e.symm.map_rel_iff.2 hab.coe
+
+@[simp] lemma copyCount_eq_zero : G.copyCount H = 0 ↔ ¬ H ⊑ G := by
```

Comment:
Do you want a version of this (and the below) in terms of `Free`? It'd be easy to prove, but still